app/src/org/commcare/CommCareApp.java

```diff
@@ -296,11 +299,28 @@ public AndroidCommCarePlatform getCommCarePlatform() {
         return platform;
     }
 
-    public boolean hasTrainingMenu() {
+    public boolean hasVisibleTrainingMenu() {
+        // This is the same eval context that would be used to evaluate the relevancy conditions of
+        // these menus when they're actually loaded
+        EvaluationContext ec =
+                CommCareApplication.instance().getCurrentSessionWrapper().getEvaluationContext(Menu.TRAINING_MENU_ROOT);
+
         for (Suite s : platform.getInstalledSuites()) {
-            List<Menu> trainingMenus = s.getMenusWithId(Menu.TRAINING_MENU_ROOT);
+            List<Menu> trainingMenus = s.getMenusWithRoot(Menu.TRAINING_MENU_ROOT);
             if (trainingMenus != null) {
-                return true;
+                for (Menu m : trainingMenus) {
+                    try {
+                        if (m.getMenuRelevance() == null ||
+                                FunctionUtils.toBoolean(m.getMenuRelevance().eval(ec))) {
+                            return true;
+                        }
+                    } catch (XPathSyntaxException e) {

```

Comment:
I'm not sure I'm aligned with the logic here. feels a little bit like if there's an invalid xpath expression here it's useful to fail as early as possible. Is it out of the question to let this bubble up to the point that we trigger an unrecoverable error (like we do when there's an error later in the app)? If it makes things significantly muckier we can skip it, but especially Syntax errors are "the app can't work as designed" kind of problems, so failing to load entirely seems like the most fail-fast behavior.

Reply:
I'm going to merge this as-is so it's in before QA, but if you disagree still let's discuss further.

Reply:
Quick note: None of this is blocking (which is why I marked as approved)

I think the 'be visible down the road' assumption is a bit dangerous. In my mind errors which are structural (IE: XPath syntax is just wrong) are always better to reveal earlier rather than later, since they are unrecoverable from the end-user (IE: CHW, not app builder) perspective, so revealing them in a 'blocking' way forces the app builder to not confuse a broken app with a functional one. 

Maybe the best way to frame this thought is "What is the value of allowing an app builder who has defined a broken condition to continue using the application while a portion of it is broken?" I think the downside of being blocked is lower from the downside of potentially not clicking the button and realizing that the app is broken.